x/auth/vesting/msg_server.go

```diff
@@ -99,10 +99,183 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
 	return &types.MsgCreateVestingAccountResponse{}, nil
 }
 
+// CreateClawbackVestingAccount creates a new ClawbackVestingAccount, or merges a grant into an existing one.
 func (s msgServer) CreateClawbackVestingAccount(goCtx context.Context, msg *types.MsgCreateClawbackVestingAccount) (*types.MsgCreateClawbackVestingAccountResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	ak := s.AccountKeeper
+	bk := s.BankKeeper
+
+	from, err := sdk.AccAddressFromBech32(msg.FromAddress)
+	if err != nil {
+		return nil, err
+	}
+	to, err := sdk.AccAddressFromBech32(msg.ToAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	if bk.BlockedAddr(to) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", msg.ToAddress)
+	}
+
+	vestingCoins := sdk.NewCoins()
+	for _, period := range msg.VestingPeriods {
+		vestingCoins = vestingCoins.Add(period.Amount...)
+	}
+
+	lockupCoins := sdk.NewCoins()
+	for _, period := range msg.LockupPeriods {
+		lockupCoins = lockupCoins.Add(period.Amount...)
+	}
+
+	// if lockup absent, default to an instant unlock schedule
+	lockupPeriods := msg.LockupPeriods
+	vestingPeriods := msg.VestingPeriods
+	if !vestingCoins.IsZero() && len(msg.LockupPeriods) == 0 {
+		lockupPeriods = []types.Period{
```

Comment:
Lockup Period and vesting Periods are from two different files. It's totally possible that lockupPeriod is absent but vestingCoins are present. As comment mentions, it would be the case of having no lockup period, vesting period starting right away